src/main/java/com/musala/atmosphere/client/websocket/ClientDispatcher.java

```diff
@@ -0,0 +1,308 @@
+     * console during a test execution).
+     * 
+     * @param deviceId
+     *        - identifier of a device
+     * @param invocationPasskey
+     * @param action
+     *        - {@link MessageAction message action}
+     * @param args
+     *        - the arguments of the request
+     * @return the result of the action sent from the Agent
+     * @throws CommandFailedException
+     *         - when failed to send the request action to the Server
+     */
+    public void routeAsync(String deviceId, long invocationPasskey, RoutingAction action, Object[] args)
+        throws ServerConnectionFailedException {
+        RequestMessage request = buildRequest(deviceId, invocationPasskey, action, args);
+        request.setAsync(true);
+
+        try {
+            Future<Void> future = session.getAsyncRemote().sendText(jsonUtil.serialize(request));
+            future.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new ServerConnectionFailedException();
+        }
+    }
+
+    private RequestMessage buildRequest(String deviceId, long invocationPasskey, RoutingAction action, Object[] args) {
+        RequestMessage requestMessage = new RequestMessage(MessageAction.ROUTING_ACTION, action, args);
+        requestMessage.setDeviceId(deviceId);
+        // The session identifier must be unique not only per WebSocket Session but and per action in case in multiple
+        // threads scenario. Otherwise the WebSocketCommunicationManager can return a response from another action.
+        String sessionIdentifier = this.session.getId() + "_" + action.toString();
+        requestMessage.setSessionId(sessionIdentifier);
+        requestMessage.setPasskey(invocationPasskey);
+
+        return requestMessage;
+    }
+
+    /**
+     * TODO: Consider to migrate this wait logic on the Server.
+     *
+     * Gets a {@link DeviceAllocationInformation} instance with the given {@link DeviceSelector device characteristics}.
+     *
+     * @param deviceSelector
+     *        - required {@link DeviceSelector parameters} needed to construct new {@link DeviceAllocationInformation}
+     *        instance.
+     * @return a {@link DeviceAllocationInformation} instance with the given device selector.
+     * @throws NoAvailableDeviceFoundException
+     *         - when there is no available device
+     */
+    public DeviceAllocationInformation getDeviceDescriptor(DeviceSelector deviceSelector,
+                                                           int allocateDeviceRetryCount) {
+        while (allocateDeviceRetryCount > 0) {
+            try {
+                RequestMessage request = new RequestMessage(MessageAction.DEVICE_ALLOCATION_INFORMATION,
+                                                            deviceSelector);
+                request.setSessionId(session.getId());
```

Comment:
place this line in the sendRequest method. You either way pass in the session there and this line repeats every time